test/test_torch.py

```diff
@@ -2881,6 +2881,34 @@ def test_deepcopy_scalar(self, device, dtype):
         self.assertEqual(a.size(), deepcopy(a).size())
         self.assertEqual(a, deepcopy(a))
 
+    @dtypes(*torch.testing.get_all_dtypes(include_bool=False, include_complex=False))
+    def test_div_rounding_modes(self, device, dtype):
+        if dtype.is_floating_point:
+            a = 10 * torch.randn(100, device=device, dtype=dtype)
+            # Avoid division by zero so we can test (a / b) * b == a
+            b = 0.1 + 2 * torch.rand(100, device=device, dtype=dtype)
+        else:
+            info = torch.iinfo(dtype)
+            low, high = info.min, info.max
+            a = torch.randint(low, high, (100,), device=device, dtype=dtype)
+            b = torch.randint(1, high, (100,), device=device, dtype=dtype)
+
+        d_true = torch.divide(a, b, rounding_mode='true')
+        self.assertTrue(d_true.is_floating_point())
+        self.assertEqual(d_true * b, a.to(d_true.dtype))
+
+        d_floor = torch.divide(a, b, rounding_mode='floor')
+        d_trunc = torch.divide(a, b, rounding_mode='trunc')
+
+        rounding_unsupported = (
+            dtype == torch.half and device != 'cuda' or
+            dtype == torch.bfloat16 and device != 'cpu')
+        if rounding_unsupported:
+            d_true = d_true.float()
+
+        self.assertEqual(d_floor, d_true.floor().to(dtype))
```

Comment:
This is good but we should explicitly compare with NumPy's floor divide and include adversarial values using two arrays that pair numbers like inf, nan, 0, -pi, pi, etc.

Reply:
I'm adding the test and it has revealed something interesting. Python floor division isn't just `floor(a / b)`, it's more like `floor((a - fmod(a, b)) / b)` which has some weird consequences:
```python
>>> n, d = 10, 0.1; n/d, (n/d).hex(), (n//d), math.fmod(n, d).hex
(100.0, '0x1.9000000000000p+6', 99.0, 0x1.9999999999972p-4)
```
Here, `10 // 0.1 != floor(10 / 0.1)` because the result of `fmod` is isn't exact. Even though the floating point division gave the exact result. This also means `inf` doesn't propagate the same, since `fmod` results in nan when `inf` is involved.

Looking at the history, I think it's done this way to maintain the relationship that `x == (x // y) * y + (x % y)`.

Reply:
This is extremely interesting. So you're saying np.floor_divide != Python's floor division? Also we should compare against Python's floor division, too.

Reply:
Oh no, numpy uses the same definition as CPython, compare:
https://github.com/numpy/numpy/blob/ce82028409c1147a6df62d8f7437e0a9262ee2b7/numpy/core/src/npymath/npy_math_internal.h.src#L714

https://github.com/python/cpython/blob/ace008c531dd685a30c1dd68f9b5ba35f20171cf/Objects/floatobject.c#L636

I'm just highlighting some non-obvious behavior I observed and will need to change the implementation to match.

Reply:
Thanks for being so thorough. We should make this clear in our documentation, too.